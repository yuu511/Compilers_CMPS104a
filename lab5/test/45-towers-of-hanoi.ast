ROOT "<<ROOT>>" 0.0.0 {0} 
FUNCTION "" 6.22.1 {0} void function 
TYPE_ID "" 6.22.1 {0} 
VOID "void" 6.22.1 {0} 
IDENT "fail" 6.22.6 {0} 
PARAM "(" 6.22.11 {0} 
TYPE_ID "" 6.22.12 {1} string variable param lval 
STRING "string" 6.22.12 {0} 
IDENT "expr" 6.22.19 {0} 
TYPE_ID "" 6.22.25 {1} string variable param lval 
STRING "string" 6.22.25 {0} 
IDENT "file" 6.22.32 {0} 
TYPE_ID "" 6.22.38 {1} int variable param lval 
INT "int" 6.22.38 {0} 
IDENT "line" 6.22.42 {0} 
FUNCTION "" 6.24.1 {1} void function 
TYPE_ID "" 6.24.1 {0} 
VOID "void" 6.24.1 {0} 
IDENT "putchr" 6.24.6 {0} 
PARAM "(" 6.24.13 {0} 
TYPE_ID "" 6.24.14 {2} int variable param lval 
INT "int" 6.24.14 {0} 
IDENT "chr" 6.24.18 {0} 
FUNCTION "" 6.25.1 {2} void function 
TYPE_ID "" 6.25.1 {0} 
VOID "void" 6.25.1 {0} 
IDENT "putint" 6.25.6 {0} 
PARAM "(" 6.25.13 {0} 
TYPE_ID "" 6.25.14 {3} int variable param lval 
INT "int" 6.25.14 {0} 
IDENT "num" 6.25.18 {0} 
FUNCTION "" 6.26.1 {3} void function 
TYPE_ID "" 6.26.1 {0} 
VOID "void" 6.26.1 {0} 
IDENT "putstr" 6.26.6 {0} 
PARAM "(" 6.26.13 {0} 
TYPE_ID "" 6.26.14 {4} string variable param lval 
STRING "string" 6.26.14 {0} 
IDENT "str" 6.26.21 {0} 
FUNCTION "" 6.28.1 {4} int function 
TYPE_ID "" 6.28.1 {0} 
INT "int" 6.28.1 {0} 
IDENT "getchr" 6.28.5 {0} 
PARAM "(" 6.28.11 {0} 
FUNCTION "" 6.29.1 {5} string function 
TYPE_ID "" 6.29.1 {0} 
STRING "string" 6.29.1 {0} 
IDENT "getstr" 6.29.8 {0} 
PARAM "(" 6.29.14 {0} 
FUNCTION "" 6.30.1 {6} string function 
TYPE_ID "" 6.30.1 {0} 
STRING "string" 6.30.1 {0} 
IDENT "getln" 6.30.8 {0} 
PARAM "(" 6.30.13 {0} 
FUNCTION "" 7.5.1 {7} void function 
TYPE_ID "" 7.5.1 {0} 
VOID "void" 7.5.1 {0} 
IDENT "move" 7.5.6 {0} 
PARAM "(" 7.5.11 {0} 
TYPE_ID "" 7.5.12 {8} string variable param lval 
STRING "string" 7.5.12 {0} 
IDENT "src" 7.5.19 {0} 
TYPE_ID "" 7.5.24 {8} string variable param lval 
STRING "string" 7.5.24 {0} 
IDENT "dst" 7.5.31 {0} 
BLOCK "{" 7.5.36 {8} 
CALL "(" 7.6.11 {3} void function (6.26.1)
IDENT "putstr" 7.6.4 {0} 
STRINGCON ""Move a disk from the "" 7.6.12 {8} string const 
CALL "(" 7.7.11 {3} void function (6.26.1)
IDENT "putstr" 7.7.4 {0} 
IDENT "src" 7.7.12 {8} string variable param lval (7.5.12)
CALL "(" 7.8.11 {3} void function (6.26.1)
IDENT "putstr" 7.8.4 {0} 
STRINGCON "" to the "" 7.8.12 {8} string const 
CALL "(" 7.9.11 {3} void function (6.26.1)
IDENT "putstr" 7.9.4 {0} 
IDENT "dst" 7.9.12 {8} string variable param lval (7.5.24)
CALL "(" 7.10.11 {3} void function (6.26.1)
IDENT "putstr" 7.10.4 {0} 
STRINGCON "".\n"" 7.10.12 {8} string const 
FUNCTION "" 7.13.1 {8} void function 
TYPE_ID "" 7.13.1 {0} 
VOID "void" 7.13.1 {0} 
IDENT "towers" 7.13.6 {0} 
PARAM "(" 7.13.13 {0} 
TYPE_ID "" 7.13.14 {9} int variable param lval 
INT "int" 7.13.14 {0} 
IDENT "ndisks" 7.13.18 {0} 
TYPE_ID "" 7.13.26 {9} string variable param lval 
STRING "string" 7.13.26 {0} 
IDENT "src" 7.13.33 {0} 
TYPE_ID "" 7.13.38 {9} string variable param lval 
STRING "string" 7.13.38 {0} 
IDENT "tmp" 7.13.45 {0} 
TYPE_ID "" 7.13.50 {9} string variable param lval 
STRING "string" 7.13.50 {0} 
IDENT "dst" 7.13.57 {0} 
BLOCK "{" 7.13.62 {9} 
IF "if" 7.14.4 {9} 
LT "<" 7.14.15 {9} int vreg 
IDENT "ndisks" 7.14.8 {9} int variable param lval (7.13.14)
INTCON "1" 7.14.17 {9} int const 
RETURN "return" 7.14.20 {0} 
CALL "(" 7.15.11 {8} void function (7.13.1)
IDENT "towers" 7.15.4 {0} 
'-' "-" 7.15.19 {9} int vreg 
IDENT "ndisks" 7.15.12 {9} int variable param lval (7.13.14)
INTCON "1" 7.15.21 {9} int const 
IDENT "src" 7.15.24 {9} string variable param lval (7.13.26)
IDENT "dst" 7.15.29 {9} string variable param lval (7.13.50)
IDENT "tmp" 7.15.34 {9} string variable param lval (7.13.38)
CALL "(" 7.16.9 {7} void function (7.5.1)
IDENT "move" 7.16.4 {0} 
IDENT "src" 7.16.10 {9} string variable param lval (7.13.26)
IDENT "dst" 7.16.15 {9} string variable param lval (7.13.50)
CALL "(" 7.17.11 {8} void function (7.13.1)
IDENT "towers" 7.17.4 {0} 
'-' "-" 7.17.19 {9} int vreg 
IDENT "ndisks" 7.17.12 {9} int variable param lval (7.13.14)
INTCON "1" 7.17.21 {9} int const 
IDENT "tmp" 7.17.24 {9} string variable param lval (7.13.38)
IDENT "src" 7.17.29 {9} string variable param lval (7.13.26)
IDENT "dst" 7.17.34 {9} string variable param lval (7.13.50)
FUNCTION "" 7.20.1 {9} int function 
TYPE_ID "" 7.20.1 {0} 
INT "int" 7.20.1 {0} 
IDENT "main" 7.20.5 {0} 
PARAM "(" 7.20.10 {0} 
TYPE_ID "" 7.20.11 {10} int variable param lval 
INT "int" 7.20.11 {0} 
IDENT "argc" 7.20.15 {0} 
TYPE_ID "" 7.20.21 {10} string array variable param lval 
ARRAY "array" 7.20.21 {0} 
STRING "string" 7.20.27 {0} 
IDENT "argv" 7.20.35 {0} 
BLOCK "{" 7.20.41 {10} 
BLOCK "{" 7.21.4 {10} 
IF "if" 7.21.5 {10} 
NOT "not" 7.21.9 {10} int vreg 
EQ "==" 7.21.19 {10} int vreg 
IDENT "argc" 7.21.14 {10} int variable param lval (7.20.11)
INTCON "2" 7.21.22 {10} int const 
CALL "(" 7.21.31 {0} void function (6.22.1)
IDENT "fail" 7.21.26 {0} 
STRINGCON ""argc == 2"" 7.21.32 {10} string const 
STRINGCON ""45-towers-of-hanoi.oc"" 7.21.45 {10} string const 
INTCON "21" 7.21.70 {10} int const 
';' ";" 7.21.75 {0} 
TYPE_ID "" 7.22.4 {10} int variable local lval vreg (7.22.4)
INT "int" 7.22.4 {0} 
IDENT "count" 7.22.8 {0} 
'-' "-" 7.22.27 {10} int variable local lval vreg 
INDEX "[" 7.22.23 {10} int lval vaddr 
INDEX "[" 7.22.20 {10} string lval vaddr 
IDENT "argv" 7.22.16 {10} string array variable param lval (7.20.21)
INTCON "1" 7.22.21 {10} int const 
INTCON "0" 7.22.24 {10} int const 
CHARCON "'0'" 7.22.29 {10} int const 
BLOCK "{" 7.23.4 {10} 
IF "if" 7.23.5 {10} 
NOT "not" 7.23.9 {10} int vreg 
GT ">" 7.23.20 {10} int vreg 
IDENT "count" 7.23.14 {10} int variable local lval (7.22.4)
INTCON "0" 7.23.22 {10} int const 
CALL "(" 7.23.31 {0} void function (6.22.1)
IDENT "fail" 7.23.26 {0} 
STRINGCON ""count > 0"" 7.23.32 {10} string const 
STRINGCON ""45-towers-of-hanoi.oc"" 7.23.45 {10} string const 
INTCON "23" 7.23.70 {10} int const 
';' ";" 7.23.75 {0} 
BLOCK "{" 7.24.4 {10} 
IF "if" 7.24.5 {10} 
NOT "not" 7.24.9 {10} int vreg 
LT "<" 7.24.20 {10} int vreg 
IDENT "count" 7.24.14 {10} int variable local lval (7.22.4)
INTCON "9" 7.24.22 {10} int const 
CALL "(" 7.24.31 {0} void function (6.22.1)
IDENT "fail" 7.24.26 {0} 
STRINGCON ""count < 9"" 7.24.32 {10} string const 
STRINGCON ""45-towers-of-hanoi.oc"" 7.24.45 {10} string const 
INTCON "24" 7.24.70 {10} int const 
';' ";" 7.24.75 {0} 
CALL "(" 7.25.11 {3} void function (6.26.1)
IDENT "putstr" 7.25.4 {0} 
INDEX "[" 7.25.16 {10} string lval vaddr 
IDENT "argv" 7.25.12 {10} string array variable param lval (7.20.21)
INTCON "0" 7.25.17 {10} int const 
CALL "(" 7.26.11 {3} void function (6.26.1)
IDENT "putstr" 7.26.4 {0} 
STRINGCON "": "" 7.26.12 {10} string const 
CALL "(" 7.27.11 {2} void function (6.25.1)
IDENT "putint" 7.27.4 {0} 
IDENT "count" 7.27.12 {10} int variable local lval (7.22.4)
CALL "(" 7.28.11 {3} void function (6.26.1)
IDENT "putstr" 7.28.4 {0} 
STRINGCON "" disks\n"" 7.28.12 {10} string const 
CALL "(" 7.29.11 {8} void function (7.13.1)
IDENT "towers" 7.29.4 {0} 
IDENT "count" 7.29.12 {10} int variable local lval (7.22.4)
STRINGCON ""source"" 7.29.19 {10} string const 
STRINGCON ""temporary"" 7.29.29 {10} string const 
STRINGCON ""distination"" 7.29.42 {10} string const 
RETURN "return" 7.30.4 {10} 
INTCON "0" 7.30.11 {10} int const 
