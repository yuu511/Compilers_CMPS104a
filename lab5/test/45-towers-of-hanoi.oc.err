3ac: non-recognized expression parsed:';' 
 3ac: warning: expr parsed without assignment: ;:
3ac: invalid statement passed :;
3ac: non-recognized expression parsed:';' 
 3ac: warning: expr parsed without assignment: ;:
3ac: invalid statement passed :;
3ac: non-recognized expression parsed:';' 
 3ac: warning: expr parsed without assignment: ;:
3ac: invalid statement passed :;
NUMBER OF ERRORS REPORTED: 6
