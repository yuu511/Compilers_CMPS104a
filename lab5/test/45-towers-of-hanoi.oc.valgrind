==6648== Memcheck, a memory error detector
==6648== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==6648== Using Valgrind-3.13.0 and LibVEX; rerun with -h for copyright info
==6648== Command: ../oc 45-towers-of-hanoi.oc
==6648== 
==6648== 
==6648== HEAP SUMMARY:
==6648==     in use at exit: 0 bytes in 0 blocks
==6648==   total heap usage: 1,124 allocs, 1,124 frees, 184,624 bytes allocated
==6648== 
==6648== All heap blocks were freed -- no leaks are possible
==6648== 
==6648== For counts of detected and suppressed errors, rerun with: -v
==6648== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
