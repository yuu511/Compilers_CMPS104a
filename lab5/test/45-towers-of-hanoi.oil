move:      .function 
           .param ptr src
           .param ptr dst
           call putstr("Move a disk from the ")  
           call putstr(src)  
           call putstr(" to the ")  
           call putstr(dst)  
           call putstr(".\n")  
           return
           .end
towers:    .function 
           .param int ndisks
           .param ptr src
           .param ptr tmp
           .param ptr dst
.if0:      $t0:i = ndisks < 1
           goto .fi0 if $t0:i
.th0:      return 
.fi0:
           $t1:i = ndisks - 1
           call towers($t1:i,src,dst,tmp)  
           call move(src,dst)  
           $t2:i = ndisks - 1
           call towers($t2:i,tmp,src,dst)  
           return
           .end
main:      .function int 
           .param int argc
           .param ptr argv
           .local int count
.if1:      $t1:i = argc == 2
           $t0:i = not $t1:i  
           goto .fi1 if $t0:i
.th1:      call fail("argc == 2","45-towers-of-hanoi.oc",21)  
.fi1:
           $t4:c = argv[1 * :c]  
           $t3:i = $t4:c[0 * :i]  
           $t2:i = $t3:i - '0'
           count = $t2:i  
.if2:      $t6:i = count > 0
           $t5:i = not $t6:i  
           goto .fi2 if $t5:i
.th2:      call fail("count > 0","45-towers-of-hanoi.oc",23)  
.fi2:
.if3:      $t8:i = count < 9
           $t7:i = not $t8:i  
           goto .fi3 if $t7:i
.th3:      call fail("count < 9","45-towers-of-hanoi.oc",24)  
.fi3:
           $t9:c = argv[0 * :c]  
           call putstr($t9:c)  
           call putstr(": ")  
           call putint(count)  
           call putstr(" disks\n")  
           call towers(count,"source","temporary","distination")  
           return 0
           return
           .end
