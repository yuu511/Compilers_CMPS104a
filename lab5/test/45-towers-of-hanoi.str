string_set[   0]:   13969197794603411352 0x55d6c202e148->""45-towers-of-hanoi.oc""
                    7296690267450279874 0x55d6c2028d58->"chr"
                   13348923631108115385 0x55d6c2029718->"putstr"
string_set[   5]:   18019194267330817565 0x55d6c202f608->""count < 9""
                   12679825168387489523 0x55d6c2029278->"num"
string_set[   6]:    5720712447182791414 0x55d6c202e858->"'0'"
                    1782657156151550167 0x55d6c202e358->"count"
                    9317282556992397945 0x55d6c202bf98->"if"
string_set[   7]:   18335579659859433076 0x55d6c202fba8->"": ""
                    2297668033614959926 0x55d6c202e6f8->"0"
                    6814028903552918164 0x55d6c202d648->"array"
                   13258891007726008550 0x55d6c202a938->"dst"
string_set[   9]:   11100957978370531577 0x55d6c202a6d8->"src"
string_set[  10]:   11597697714117577063 0x55d6c202e398->"="
                    1583576014053361350 0x55d6c20298d8->"str"
                   13497578857207247412 0x55d6c2028318->"string"
string_set[  11]:   11608465765758206812 0x55d6c2030468->""distination""
                    1396094175042499165 0x55d6c202c158->"<"
string_set[  13]:   12868445110721718657 0x55d6c2028978->")"
string_set[  14]:   14168603451105147218 0x55d6c202f0c8->"23"
string_set[  15]:   11299632471807205110 0x55d6c20302b8->""source""
string_set[  16]:   11911796782241914524 0x55d6c2027088->"void"
                   10159970873491820195 0x55d6c202c218->"1"
string_set[  18]:   13581035385482130682 0x55d6c202e4d8->"["
string_set[  19]:    5344662657899890615 0x55d6c2028a38->";"
string_set[  20]:   11868120863286193964 0x55d6c202d2c8->"main"
                   15816506932648033945 0x55d6c2028258->"("
string_set[  25]:   15225576330967767377 0x55d6c202eeb8->""count > 0""
                   11166701219533395393 0x55d6c2029ef8->"getstr"
string_set[  29]:   14635384904416252381 0x55d6c202c378->"return"
                   11152315395316048902 0x55d6c202b648->"towers"
string_set[  34]:    7211130878751977441 0x55d6c2028b98->"putchr"
string_set[  35]:   14736352883882276958 0x55d6c202dd28->"=="
string_set[  36]:   11809324432133475174 0x55d6c202a518->"move"
string_set[  37]:   12291392672316406109 0x55d6c2028898->"line"
string_set[  39]:   18413295556186398326 0x55d6c20290b8->"putint"
string_set[  43]:    7372855250431704371 0x55d6c202c758->"-"
string_set[  44]:      21606623943142546 0x55d6c202aab8->"{"
                    8559740906699418289 0x55d6c202ac78->""Move a disk from the ""
string_set[  45]:    6142509188972423790 0x55d6c2028498->""
string_set[  47]:   16741722213504744880 0x55d6c202b398->"".\n""
                   13686721134187162920 0x55d6c202b4c8->"}"
string_set[  51]:    3701972582333163920 0x55d6c202d488->"argc"
string_set[  53]:   14710271594181063355 0x55d6c202ffb8->"" disks\n""
                   12051031859636400188 0x55d6c202e618->"]"
                   11001320821098761033 0x55d6c202b038->"" to the ""
string_set[  55]:    6623666755989985924 0x55d6c202f408->"9"
string_set[  57]:    9386170080689500517 0x55d6c20283d8->"expr"
string_set[  58]:    4017852018407011111 0x55d6c2028658->"file"
                    3729804957429652673 0x55d6c2021d08->"<<ROOT>>"
string_set[  59]:    2899015600721308611 0x55d6c202e238->"21"
                   16579916199511856337 0x55d6c202a1f8->"getln"
string_set[  61]:    4551451650890805270 0x55d6c202dde8->"2"
                   14849783028996024063 0x55d6c202db68->"not"
                   14643508047410943861 0x55d6c202d868->"argv"
string_set[  63]:   15577135225310562476 0x55d6c20303a8->""temporary""
                    2749921394955183836 0x55d6c202f808->"24"
                    9611428142580219387 0x55d6c2028158->"fail"
string_set[  64]:   12597418120160561823 0x55d6c20287d8->"int"
string_set[  70]:    9201800968376702500 0x55d6c2029bf8->"getchr"
string_set[  71]:   14526079891305879776 0x55d6c202d808->">"
                    5026105222461765741 0x55d6c202b808->"ndisks"
string_set[  73]:     232592375681467293 0x55d6c202bc08->"tmp"
string_set[  77]:    9881412649056685758 0x55d6c202dfc8->""argc == 2""
string_set[  78]:    4706206694200543916 0x55d6c2028198->","
load_factor = 0.823
bucket_count = 79
max_bucket_size = 4
