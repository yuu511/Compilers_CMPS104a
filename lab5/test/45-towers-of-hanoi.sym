towers (7.13.1) {8} void function 
   ndisks (7.13.14) {9} int variable param lval 0
   src (7.13.26) {9} string variable param lval 1
   tmp (7.13.38) {9} string variable param lval 2
   dst (7.13.50) {9} string variable param lval 3

main (7.20.1) {9} int function 
   argc (7.20.11) {10} int variable param lval 0
   argv (7.20.21) {10} string array variable param lval 1
   count (7.22.4) {10} int variable local lval 2

move (7.5.1) {7} void function 
   src (7.5.12) {8} string variable param lval 0
   dst (7.5.24) {8} string variable param lval 1

getln (6.30.1) {6} string function 

putchr (6.24.1) {1} void function 

putstr (6.26.1) {3} void function 

getchr (6.28.1) {4} int function 

getstr (6.29.1) {5} string function 

fail (6.22.1) {0} void function 

putint (6.25.1) {2} void function 

