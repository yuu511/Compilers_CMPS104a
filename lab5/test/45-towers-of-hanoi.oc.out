CONDITIONAL IF STATEMENT SYMBOL TOK_LT 
CONDITIONAL IF STATEMENT SYMBOL TOK_NOT 
CONDITIONAL IF STATEMENT SYMBOL TOK_NOT 
CONDITIONAL IF STATEMENT SYMBOL TOK_NOT 
